### PR TITLE
Add gRPC proto def for rate-limit xDS Config communication

### PR DESCRIPTION
**Pull request by renuka-fernando**
### Description

Proto for xDS communication as in the discussion https://github.com/envoyproxy/ratelimit/issues/201#issuecomment-1277521988.
xDS implementation: https://github.com/envoyproxy/ratelimit/pull/373

### CODE REVIEW

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+  uint32 requests_per_unit = 2;
+  bool unlimited = 3;
+  string name = 4;
+  repeated RateLimitReplace replaces = 5;
+}
+
+// Rate-limit replace model
+message RateLimitReplace {
+  string name = 1;
+}
```
Comment (mattklein123) [inline api/ratelimit/config/ratelimit/v3/rls_conf.proto:39]: What is this?

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto [hunk 2]
```diff
@@ -0,0 +1,39 @@
+syntax = "proto3";
+
+package ratelimit.config.ratelimit.v3;
+
+option java_package = "io.envoyproxy.ratelimit.config.ratelimit.v3";
+option java_outer_classname = "RlsConfigProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/ratelimit/config/ratelimit/v3;ratelimitv3";
+
+// [#protodoc-title: Rate Limit Config Discovery Service (RLS Conf DS)]
```
Comment (mattklein123) [inline api/ratelimit/config/ratelimit/v3/rls_conf.proto:10]: Remove and replace with more comments.

**Review by mattklein123 — CHANGES_REQUESTED** [review]
Thanks for working on this. Can you add more comments so it's easier to understand how this is going to be used in the context of ADS? Thank you.

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto (hunk 2 above)
Comment (renuka-fernando) [reply api/ratelimit/config/ratelimit/v3/rls_conf.proto]: Fixed with https://github.com/envoyproxy/ratelimit/pull/371/commits/0f1a7494234a24262e4f0c5bc00d614f934a8ae3

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto (hunk 1 above)
Comment (renuka-fernando) [reply api/ratelimit/config/ratelimit/v3/rls_conf.proto]: Described this in the comments of the proto file. This is for the "replace" feature.
Feature https://github.com/envoyproxy/ratelimit/tree/0b2f4d5fb04bf55e1873e2c5e2bb28da67c0643f#replaces
Example: https://github.com/envoyproxy/ratelimit/tree/0b2f4d5fb04bf55e1873e2c5e2bb28da67c0643f#example-7

**Comment by renuka-fernando** [comment]
I had a discussion with @alecholmez about adding this along with ADS. We can use ADS for this and there is no ADS client in go-control-plane hence, I will add an xDS client in this repo.

For writing the xDS server we can use go-control-plane and use Snapshots to update the rate limits. This PR https://github.com/envoyproxy/go-control-plane/pull/598 will register the rate limit config as a known type in go-control-plane.

**Comment by mattklein123** [comment]
> I had a discussion with @alecholmez about adding this along with ADS. We can use ADS for this and there is no ADS client in go-control-plane hence, I will add an xDS client in this repo.

Please don't add the ADS client in this repo. I think it belongs in GCP so it can be used in other places.

**Comment by renuka-fernando** [comment]
> > I had a discussion with @alecholmez about adding this along with ADS. We can use ADS for this and there is no ADS client in go-control-plane hence, I will add an xDS client in this repo.
> 
> Please don't add the ADS client in this repo. I think it belongs in GCP so it can be used in other places.

Used the ADS streaming client https://github.com/envoyproxy/go-control-plane/blob/129d0216a4efb985b22349fb45c1444d4f0e7900/envoy/service/discovery/v3/ads.pb.go#L228
Please find the xDS implementation in https://github.com/envoyproxy/ratelimit/pull/373

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto [hunk 3]
```diff
@@ -0,0 +1,86 @@
+syntax = "proto3";
+
+package ratelimit.config.ratelimit.v3;
+
+option java_package = "io.envoyproxy.ratelimit.config.ratelimit.v3";
+option java_outer_classname = "RlsConfigProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/ratelimit/config/ratelimit/v3;ratelimitv3";
+
+// [#protodoc-title: Rate limit service configuration]
```
Comment (mattklein123) [inline api/ratelimit/config/ratelimit/v3/rls_conf.proto:10]: Can you add more comments on how this is going to be used with ADS? I assume the stream will be a series of RateLimitConfig either in SoTW or delta form?

**Review by mattklein123 — CHANGES_REQUESTED** [review]

File: api/ratelimit/config/ratelimit/v3/rls_conf.proto (hunk 3 above)
Comment (renuka-fernando) [reply api/ratelimit/config/ratelimit/v3/rls_conf.proto:10]: Added more comments with https://github.com/envoyproxy/ratelimit/pull/371/commits/d7b55467c7d1d7dc89d1bc540e4b79b37884631a

**Comment by renuka-fernando** [comment]
Could you please retrigger the build? This is an intermitten issue, which will be fixed with https://github.com/envoyproxy/ratelimit/pull/370

**Review by mattklein123 — COMMENTED** [review]
This looks fine but one thing I'm wondering is can we share the protos from xds/envoy for some of the core structures?

**Comment by kkcmadhu** [comment]
this app prodvides an http endpoint /json to get the current rate limit as jason.
A HTTP endpoint to add /modify the current config will be great too..

irrespective of  GRPC or HTTP, do we think we will need some kind of role support to control who can modify this config?

**Comment by renuka-fernando** [comment]
Hi @mattklein123,

ADS Client implementation is merged https://github.com/envoyproxy/go-control-plane/pull/604 and updated the go-control-plane version to reflect the client implementation. Could you please review this PR?

Thanks.

**Comment by renuka-fernando** [comment]
Closing this PR. The proto file is included in https://github.com/envoyproxy/ratelimit/pull/373 as a single commit.

**Pull request closed without merging**